### PR TITLE
nginxQuic package and boringssl update

### PR DESCRIPTION
**Pull request by mkg20001**
###### Motivation for this change

Adds nginxQuic from https://quic.nginx.org/ and an option to use it in vhosts (vhost.http3 = true;)

Also updates boringssl to a newer version that supports h3 (and some changes as go modules need to be fetched now)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-servers/nginx/vhost-options.nix [hunk 1]
```diff
@@ -151,6 +151,18 @@ with lib;
...
     };
 
+    http3 = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Whether to enable HTTP 3.
+        This requires <literal>services.nginx.package = pkgs.nginxQuic;</literal>
+        Note that HTTP 3 support is experimental and
+        *not* recommended for production.
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/nginx/vhost-options.nix:161]:
```suggestion
        *not* yet recommended for production.
```

File: nixos/modules/services/web-servers/nginx/vhost-options.nix [hunk 2]
```diff
@@ -151,6 +151,18 @@ with lib;
       '';
     };
 
+    http3 = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Whether to enable HTTP 3.
+        This requires <literal>services.nginx.package = pkgs.nginxQuic;</literal>
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/nginx/vhost-options.nix:159]:
```suggestion
        This requires using <literal>pkgs.nginxQuic</literal> package
        which can be achived by setting <literal>services.nginx.package = pkgs.nginxQuic;</literal>.
```

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 3]
```diff
@@ -249,7 +249,12 @@ let
           + optionalString (ssl && vhost.http2) "http2 "
           + optionalString vhost.default "default_server "
           + optionalString (extraParameters != []) (concatStringsSep " " extraParameters)
-          + ";";
+          + ";"
+          + (if ssl && vhost.http3 then ''
+          listen ${addr}:${toString port} http3 reuseport;  # UDP listener for QUIC+HTTP/3
+          add_header Alt-Svc 'h3=":443"';   # Advertise that HTTP/3 is available
+          add_header QUIC-Status $quic;     # Sent when QUIC was used
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-servers/nginx/default.nix:256]:
```suggestion
          # UDP listener for **QUIC+HTTP/3
          listen ${addr}:${toString port} http3 reuseport;
          # Advertise that HTTP/3 is available
          add_header Alt-Svc 'h3=":443"';
          # Sent when QUIC was used
          add_header QUIC-Status $quic;
```

File: pkgs/development/libraries/boringssl/default.nix [hunk 4]
```diff
@@ -1,22 +1,38 @@
...
+, fetchgit
+, cmake
+, ninja
+, perl
+, buildGoModule
+}:
 
 # reference: https://boringssl.googlesource.com/boringssl/+/2661/BUILDING.md
-stdenv.mkDerivation {
+# stdenv.mkDerivation {
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/boringssl/default.nix:11]:
```suggestion
```

File: pkgs/development/libraries/boringssl/default.nix [hunk 5]
```diff
@@ -1,22 +1,38 @@
...
+buildGoModule { # we need some go modules to be fetched, so we use this
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/boringssl/default.nix:12]:
```suggestion
buildGoModule {
```

File: pkgs/development/libraries/boringssl/default.nix [hunk 6]
```diff
@@ -1,22 +1,38 @@
...
-  makeFlags = [ "GOCACHE=$(TMPDIR)/go-cache" ];
+  vendorSha256 = "sha256-J+FVWVPS2TK792zbKu0Nu/iCiz5EU+Sy1qHBmogEyXA=";
+
+  preBuild = ''
+    cmakeConfigurePhase
+  '';
+
+  buildPhase = ''
+    ninjaBuildPhase
+  '';
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/boringssl/default.nix:32]: Shouldn't those be invoked automatically?

File: pkgs/top-level/all-packages.nix [hunk 7]
```diff
@@ -18665,6 +18665,15 @@ in
 
   nginx = nginxStable;
 
+  nginxQuic = callPackage ../servers/http/nginx/quic.nix {
+    withPerl = false;
+    # We don't use `with` statement here on purpose!
+    # See https://github.com/NixOS/nixpkgs/pull/10474/files#r42369334
+    modules = [ nginxModules.rtmp nginxModules.dav nginxModules.moreheaders ];
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:18672]: How should I change the quic package in the module? I would need to copy this expression and adapt it right? Maybe we should mention this.

File: pkgs/top-level/all-packages.nix [hunk 8]
```diff
@@ -18665,6 +18665,15 @@ in
 
   nginx = nginxStable;
 
+  nginxQuic = callPackage ../servers/http/nginx/quic.nix {
+    withPerl = false;
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:18669]: I am curious why perl is disabled?

**Review by SuperSandro2000 — COMMENTED** [review]
Some ideas I had while reading. Good idea.

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 119813` at dc1214dc run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages marked as broken and skipped:</summary>
  <ul>
    <li>python38Packages.arviz</li>
    <li>python38Packages.dm-sonnet</li>
    <li>python38Packages.graph_nets</li>
    <li>python38Packages.optuna</li>
    <li>python38Packages.pymc3</li>
    <li>python38Packages.rl-coach</li>
    <li>python38Packages.tensorflow-probability</li>
    <li>python39Packages.dm-sonnet</li>
    <li>python39Packages.graph_nets</li>
    <li>python39Packages.optuna</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/34f80376227b0e2087c2275d706a53cd">boringssl</a></li>
    <li><a href="https://gist.github.com/16f30d61297b3f85bb670d5cd605b6e5">nginxQuic</a></li>
  </ul>
</details>
<details>
  <summary>16 packages skipped due to time constraints:</summary>
  <ul>
    <li>gpt2tc</li>
    <li>python38Packages.baselines</li>
    <li>python38Packages.edward</li>
    <li>python38Packages.mask-rcnn</li>
    <li>python38Packages.scikit-tda</li>
    <li>python38Packages.tensorflow (python38Packages.tensorflow-build ,python38Packages.tensorflow-build_2 ,python38Packages.tensorflowWithoutCuda ,python38Packages.tensorflow_2)</li>
    <li>python38Packages.tensorflowWithCuda</li>
    <li>python38Packages.tflearn</li>
    <li>python38Packages.umap-learn</li>
    <li>python39Packages.baselines</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`buildPhase` should probably contain `runHook preBuild`.

Near pkgs/development/go-modules/generic/default.nix:145:5:
```
    |
145 |     buildPhase = args.buildPhase or ''
    |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/development/go-modules/generic/default.nix:220:5:
```
    |
220 |     installPhase = args.installPhase or ''
    |     ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement.

File: pkgs/development/libraries/boringssl/default.nix (hunk 6 above)
Comment (mkg20001) [reply pkgs/development/libraries/boringssl/default.nix:33]: buildGoModule overwrites them

File: pkgs/top-level/all-packages.nix (hunk 7 above)
Comment (mkg20001) [reply pkgs/top-level/all-packages.nix:18686]: Quic currently is part of the HG repository and gets pulled with nginx-quic, there's no seperate module file

File: pkgs/top-level/all-packages.nix (hunk 8 above)
Comment (mkg20001) [reply pkgs/top-level/all-packages.nix:18683]: Copied it from below (nginxStable), no idea

File: pkgs/top-level/all-packages.nix (hunk 7 above)
Comment (mkg20001) [reply pkgs/top-level/all-packages.nix:18686]: There's also quiche as alternative quic server, has different options and works via a patch + a seperate rust library linked into nginx.

Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix:18686]: If I wanted to add another module this wouldn't be easily possible right now.

Comment (mkg20001) [reply pkgs/top-level/all-packages.nix:18686]: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-servers/nginx/default.nix#L404-L416 there's an option for that, also other packages do it like that, too

**Pull request merged**